File: app/checker.py
import asyncio
import random
import time
import httpx
import aiohttp
from aiohttp_socks import ProxyConnector
from .config import settings
from .db import Database

CHECK_URLS = tuple(x.strip() for x in settings.check_targets.split(",") if x.strip())


async def check_one(proxy: str, db: Database, semaphore: asyncio.Semaphore):
    async with semaphore:
        started = time.perf_counter()
        try:
            results = []
            if proxy.startswith(("socks4://", "socks5://")):
                connector = ProxyConnector.from_url(proxy)
                async with aiohttp.ClientSession(connector=connector) as client:
                    for target in random.sample(CHECK_URLS, k=len(CHECK_URLS)):
                        async with client.get(target, timeout=settings.check_timeout_seconds, allow_redirects=False) as response:
                            results.append(response.status in (200, 204, 301, 302))
            else:
                async with httpx.AsyncClient(proxy=proxy, timeout=settings.check_timeout_seconds, follow_redirects=False) as client:
                    for target in random.sample(CHECK_URLS, k=len(CHECK_URLS)):
                        response = await client.get(target, headers={"User-Agent": "proxy-pool-health-check/1.0"})
                        results.append(response.status_code in (200, 204, 301, 302))
            latency = round((time.perf_counter() - started) * 1000, 2)
            # A proxy is accepted only if it responds within the configured SLA.
            alive = all(results) and latency <= settings.check_timeout_seconds * 1000
            await db.update_check(proxy, alive, latency if alive else None)
            return {"alive": alive, "error": None}
        except asyncio.CancelledError:
            # Cancellation belongs to the caller (and must never be swallowed).
            raise
        except Exception as exc:
            # Public proxies frequently close a SOCKS handshake midway through
            # an asyncio read.  aiohttp-socks can surface that as
            # IncompleteReadError, TimeoutError, OSError, or a connector-specific
            # exception, so keep this boundary deliberately broad.  A bad proxy
            # is a normal result of a health check, not a batch failure.
            try:
                await db.update_check(proxy, False, None)
            except asyncio.CancelledError:
                raise
            except Exception as db_exc:
                return {"alive": False, "error": f"database:{type(db_exc).__name__}"}
            return {"alive": False, "error": type(exc).__name__}


async def check_all(db: Database, alive_only: bool = False):
    # Checking in bounded batches prevents thousands of public entries from
    # creating an unbounded number of tasks and file descriptors.
    proxies = await db.candidates(alive_only=alive_only)
    if settings.max_candidates_per_cycle > 0:
        proxies = proxies[: settings.max_candidates_per_cycle]
    semaphore = asyncio.Semaphore(settings.check_concurrency)
    results = await asyncio.gather(
        *(check_one(p, db, semaphore) for p in proxies),
        return_exceptions=True,
    )
    unexpected = 0
    failed = 0
    error_types = {}
    for result in results:
        if isinstance(result, asyncio.CancelledError):
            raise result
        if isinstance(result, Exception):
            # Defensive guard for future changes to check_one.  It keeps one
            # proxy task from cancelling the complete scan.
            unexpected += 1
            print(f"[check] unexpected task failure: {type(result).__name__}")
        elif not result["alive"]:
            failed += 1
            if result["error"]:
                error_types[result["error"]] = error_types.get(result["error"], 0) + 1
    return {"tested": len(proxies), "failed": failed, "unexpected": unexpected, "error_types": error_types}

File: src/checker.js
import axios from 'axios';
import { HttpProxyAgent } from 'http-proxy-agent';
import { HttpsProxyAgent } from 'https-proxy-agent';
import { SocksProxyAgent } from 'socks-proxy-agent';
import { settings } from './config.js';
import { errorCode, errorDetail, isResourceError, networkGuard } from './network.js';

function agentsFor(proxy, signal) {
  const timeout = settings.checkTimeoutSeconds * 1000;
  const normalized = proxy.startsWith('https://') ? `http://${proxy.slice(8)}` : proxy;
  if (normalized.startsWith('socks4://') || normalized.startsWith('socks5://')) {
    const agent = new SocksProxyAgent(normalized, { timeout, socketOptions: { signal } });
    return { httpAgent: agent, httpsAgent: agent };
  }
  // Agent.destroy() cannot see sockets still awaiting CONNECT. Cancel the
  // underlying net/tls socket too, not just Axios's ClientRequest.
  const options = { signal, timeout, keepAlive: false };
  return { httpAgent: new HttpProxyAgent(normalized, options), httpsAgent: new HttpsProxyAgent(normalized, options) };
}

async function probe(proxy, signal) {
  const agents = agentsFor(proxy, signal);
  try {
    for (const target of settings.checkTargets) {
      const response = await axios.get(target, {
        ...agents,
        proxy: false,
        timeout: settings.checkTimeoutSeconds * 1000,
        signal,
        maxRedirects: 0,
        maxContentLength: 64 * 1024,
        validateStatus: () => true,
        headers: { 'User-Agent': 'proxy-pool-health-check/3.0' },
      });
      if (response.status !== 204) throw Object.assign(new Error(`Expected 204, received HTTP ${response.status}`), { code: `HTTP_STATUS_${response.status}` });
    }
    return true;
  } finally {
    agents.httpAgent?.destroy?.();
    if (agents.httpsAgent !== agents.httpAgent) agents.httpsAgent?.destroy?.();
  }
}

export async function checkOne(proxy, { signal } = {}) {
  const started = performance.now();
  const timeoutMs = settings.checkTimeoutSeconds * 1000;
  const controller = new AbortController();
  const combinedSignal = signal ? AbortSignal.any([signal, controller.signal]) : controller.signal;
  const deadline = setTimeout(() => controller.abort(), timeoutMs);
  deadline.unref?.();
  try {
    combinedSignal.throwIfAborted();
    const passed = await probe(proxy, combinedSignal);
    const latency = Math.round((performance.now() - started) * 100) / 100;
    const alive = passed && latency <= timeoutMs;
    return { proxy, alive, latency: alive ? latency : null, error: alive ? null : 'TimeoutError' };
  } catch (error) {
    if (signal?.aborted) return { proxy, alive: false, latency: null, error: 'CANCELLED', cancelled: true };
    const code = errorCode(error);
    const reason = isResourceError(code) ? code : (controller.signal.aborted ? 'TimeoutError' : code);
    return { proxy, alive: false, latency: null, error: reason, detail: errorDetail(error) };
  } finally {
    clearTimeout(deadline);
    controller.abort();
  }
}

export async function checkAll(db, { aliveOnly = false, staleOnly = false,
  knownOnly = false, discoveryOnly = false, dueSeconds = settings.staleAfterSeconds,
  checker = checkOne, guard = networkGuard, scheduler, signal, persist = true } = {}) {
  const started = performance.now();
  let proxies = db.candidates({ aliveOnly, staleOnly, knownOnly, discoveryOnly, dueSeconds });
  // A discovery cap must never exclude already-working proxies from rechecks.
  if (!knownOnly && !aliveOnly && settings.maxCandidatesPerCycle > 0) proxies = proxies.slice(0, settings.maxCandidatesPerCycle);
  const summary = { tested: 0, failed: 0, succeeded: 0, deferred: 0, skipped: 0,
    unexpected: 0, errorTypes: {}, errorSamples: {} };
  let next = 0;
  let sinceSave = 0;
  const recordError = (code, detail) => {
    const key = Object.hasOwn(summary.errorTypes, code) || Object.keys(summary.errorTypes).length < 40 ? code : 'OTHER';
    summary.errorTypes[key] = (summary.errorTypes[key] || 0) + 1;
    if (detail && !summary.errorSamples[key]) summary.errorSamples[key] = detail;
  };
  const flush = () => {
    if (!persist || !db.dirty) return;
    try { db.save(); sinceSave = 0; }
    catch (error) { summary.unexpected++; recordError(`database:${errorCode(error)}`, errorDetail(error)); }
  };
  const run = async (proxy) => {
    // The discovery snapshot may be minutes old by the time a slot is free.
    if (knownOnly || discoveryOnly) {
      const row = db.queryOne('SELECT last_success FROM proxies WHERE proxy=?', [proxy]);
      if (!row || (knownOnly && !row.last_success) || (discoveryOnly && row.last_success)) return;
    }
    let result;
    try { result = await checker(proxy, { signal }); }
    catch (error) { result = { alive: false, error: errorCode(error), detail: errorDetail(error) }; }
    summary.tested++;
    if (result.cancelled || signal?.aborted) { summary.deferred++; return; }
    guard.fail(result.error, settings.checkConcurrency);
    if (result.error) recordError(result.error, result.detail);
    if (!result.alive && guard.paused()) { summary.deferred++; return; }
    summary[result.alive ? 'succeeded' : 'failed']++;
    try {
      // Publish each completion immediately. Disk exports are coalesced by the
      // service's flush loop (or every configured batch for snapshot scripts).
      db.updateChecks([[proxy, result.alive, result.latency ?? null, new Date().toISOString()]], { persist: false });
      sinceSave++;
    } catch (error) { summary.unexpected++; recordError(`database:${errorCode(error)}`, errorDetail(error)); }
    if (sinceSave >= settings.checkPersistBatchSize) flush();
  };
  const worker = async () => {
    while (!signal?.aborted && !guard.paused()) {
      const proxy = proxies[next++];
      if (proxy === undefined) return;
      if (scheduler) await scheduler.run(proxy, () => run(proxy), { priority: knownOnly ? 1 : 0, signal });
      else await run(proxy);
    }
  };
  await Promise.all(Array.from({ length: Math.min(guard.concurrency(settings.checkConcurrency), proxies.length) }, worker));
  flush();
  if (summary.succeeded > 0 && !guard.paused()) guard.succeeded(settings.checkConcurrency);
  summary.skipped = proxies.length - summary.tested;
  summary.duration_ms = Math.round(performance.now() - started);
  return summary;
}
